wrangler-core/src/main/java/io/cdap/wrangler/utils/Json2Schema.java

```diff
@@ -73,62 +73,75 @@ public Schema toSchema(String id, Row row) throws RecordConvertorException {
+        throw new RecordConvertorException(
+          String.format("Unable to convert field '%s' to basic type.", name)
+        );
       }
+    }
 
-      if (value instanceof LocalDate) {
-        Schema schema = Schema.nullableOf(Schema.of(Schema.LogicalType.DATE));
-        fields.add(Schema.Field.of(name, schema));
-      }
+    if (value instanceof BigDecimal) {
+      // TODO CDAP-15361 precision should be derived from the schema of the row.
+      return Schema.nullableOf(Schema.decimalOf(38, ((BigDecimal) value).scale()));
+    }
 
-      if (value instanceof LocalTime) {
-        Schema schema = Schema.nullableOf(Schema.of(Schema.LogicalType.TIME_MICROS));
-        fields.add(Schema.Field.of(name, schema));
-      }
+    if (value instanceof LocalDate) {
+      return Schema.nullableOf(Schema.of(Schema.LogicalType.DATE));
+    }
 
-      if (value instanceof ZonedDateTime) {
-        Schema schema = Schema.nullableOf(Schema.of(Schema.LogicalType.TIMESTAMP_MICROS));
-        fields.add(Schema.Field.of(name, schema));
-      }
+    if (value instanceof LocalTime) {
+      return Schema.nullableOf(Schema.of(Schema.LogicalType.TIME_MICROS));
+    }
 
-      // TODO - remove all the instaces of java.util.Date once all the directives support LogicalType.
-      if (value instanceof Date || value instanceof java.sql.Date || value instanceof Time
-        || value instanceof Timestamp) {
-        Schema schema = Schema.nullableOf(Schema.of(Schema.Type.LONG));
-        fields.add(Schema.Field.of(name, schema));
-      }
+    if (value instanceof ZonedDateTime) {
+      return  Schema.nullableOf(Schema.of(Schema.LogicalType.TIMESTAMP_MICROS));
+    }
 
-      if (value instanceof Map) {
-        Schema schema = Schema.nullableOf (Schema.mapOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.STRING)));
-        fields.add(Schema.Field.of(name, schema));
-      }
+    // TODO - remove all the instaces of java.util.Date once all the directives support LogicalType.
+    if (value instanceof Date || value instanceof java.sql.Date || value instanceof Time
+      || value instanceof Timestamp) {
+      return Schema.nullableOf(Schema.of(Schema.Type.LONG));
+    }
 
-      if (value instanceof JsonElement) {
-        Schema schema = toSchema(name, (JsonElement) value);
-        fields.add(Schema.Field.of(name, schema));
-      }
+    if (value instanceof Map) {
+      return Schema.nullableOf (Schema.mapOf(Schema.of(Schema.Type.STRING), Schema.of(Schema.Type.STRING)));
+    }
 
-      if (value instanceof Schema) {
-        fields.addAll(((Schema) value).getFields());
+    if (value instanceof JsonElement) {
+      return toSchema(name, (JsonElement) value);
+    }
+
+    if (value instanceof List) {
+      List<?> list = (List) value;
+      for (Object listObject : list) {
+        Schema schema = getSchema(listObject, name);
+        if (schema != null) {
+          return Schema.nullableOf(Schema.arrayOf(schema));
```

Comment:
Doesn't the BQ sink fail if the array is a nullable array? 

Generally it's best if collections themselves are not nullable, with empty collections used instead of null, but I don't know if the directives have been following that.

Reply:
yes bq sink would fail if schema for array or array component is null. However, we can modify the schema from pipeline UI to make array and its component to be non-nullable. 
Array is kept as nullable for other cases.